framework/src/reporters/PerfGraphReporter.C

```diff
@@ -0,0 +1,96 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PerfGraphReporter.h"
+
+#include "PerfGraph.h"
+
+registerMooseObject("MooseApp", PerfGraphReporter);
+
+InputParameters
+PerfGraphReporter::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Reports the full performance graph from the PerfGraph.");
+  return params;
+}
+
+PerfGraphReporter::PerfGraphReporter(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _nodes(declareValueByName<std::vector<NodeEntry>>("nodes", REPORTER_MODE_DISTRIBUTED))
+{
+}
+
+void
+PerfGraphReporter::execute()
+{
+  // PerfNodes do not have unique IDs; their ID is based on the section
+  // in which they are generated. But... we want unique IDs. Hence this map
+  std::size_t next_unique_id = 0;
+  std::map<const PerfNode *, std::size_t> unique_ids;
+  const auto unique_id = [&unique_ids, &next_unique_id](const PerfNode & node) {
+    const auto find = unique_ids.find(&node);
+    if (find != unique_ids.end())
+      return find->second;
+    unique_ids[&node] = next_unique_id;
+    return next_unique_id++;
+  };
+
+  // Action that stores the nodes information info _nodes
+  const auto add_entry = [this, &unique_id](const PerfGraph::PerfNodeInfo & info) {
+    NodeEntry entry;
+    entry.name = info.sectionInfo().name();
+    entry.ints["id"] = unique_id(info.node());
+    entry.ints["level"] = info.sectionInfo().level();
+    entry.ints["num_calls"] = info.node().numCalls();
+    entry.ints["memory"] = info.node().selfMemory();
+    entry.time = info.node().selfTimeSec();
+
+    if (info.parentNode())
+      entry.ints["parent_id"] = unique_id(*info.parentNode());
+
+    for (const auto & id_node_pair : info.node().children())
+      entry.children_ids.push_back(unique_id(*id_node_pair.second));
+
+    _nodes.push_back(entry);
```

Comment:
I still feel like there should have been a way to do this without remaking the graph again before outputting...

Also - is this serializing the tree into a vector?  Does that mean that the data doesn't come out as a hierarchical tree?

Reply:
> I still feel like there should have been a way to do this without remaking the graph again before outputting

First, doing it like this enables proper loading of old PerfGraph data on restart. Secondly, I still prefer to have the `to_json` method be simplified where it's really only converting static data. This is consistent with other Reporter actions.

> Also - is this serializing the tree into a vector? Does that mean that the data doesn't come out as a hierarchical tree?

Correct. I can do either method. I output it as a tree originally, but the JSON ended up being super non human readable with all of the indentation... it was just a personal preference. Re-building the tree in Python takes little time and the difference in LOC between rebuilding the tree and not is negligible. If you would prefer to have the tree, I can do that. I'm not set on either method.

Reply:
As for restart - we should be able to store the PerfGraph nodes directly in the restart data as well.  Shouldn't be hard actually...